test/common/TestBed.hpp

```diff
@@ -0,0 +1,126 @@
+    int                        numCollectivesInGroup; // # of collectives to execute per group call
+
+    EnvVars ev;                                       // Environment variables
+
+    // Constructor - Creates one child process per detected GPU device that waits for further commands
+    TestBed();
+
+    // Prepare TestBed for use with GPUs across multiple child processes
+    void InitComms(std::vector<std::vector<int>> const& deviceIdsPerChild, int const numCollectivesInGroup = 1);
+    // Prepare TestBed for use with GPUs on a single child process
+    void InitComms(int const numGpus, int const numCollectivesInGroup = 1);
+
+    // Set collectives arguments for specified collective / rank
+    // Setting scalarsPerRank to non-null will create custom reduction operator
+    // Using collId = -1 (default) applies settings to all collectives in group
+    // Using rank = -1 (default) applies settings to all ranks
+
+    void SetCollectiveArgs(ncclFunc_t     const funcType,
+                           ncclDataType_t const dataType,
+                           ncclRedOp_t    const redOp,
+                           int            const root,
+                           size_t         const numInputElements,
+                           size_t         const numOutputElements,
+                           int            const collId = -1,
+                           int            const rank = -1,
+                           PtrUnion       const scalarsPerRank = {nullptr},
+                           int            const scalarMode = -1);
+
+    // Allocate memory for specified collective / rank
+    // - Requires SetCollectiveArgs to have been called already
+    // Using collId = -1 (default) applies settings to all collectives in group
+    // Using rank = -1 (default) applies settings to all ranks
+    void AllocateMem(bool   const inPlace = false,
+                     bool   const useManagedMem = false,
+                     int    const collId = -1,
+                     int    const rank = -1);
+
+    // Initialize input and compute expected results
+    // - requires that SetCollectiveArgs and AllocateMemory have already been called
+    // Setting collId to -1 applies settings to all collectives in group
+    // Setting rank to -1 applies settings to all ranks
+    // Setting prepDataFunc to nullptr uses the default fill pattern routine
+    void PrepareData(int const collId = -1,
+                     int const rank = -1,
+                     CollFuncPtr const prepDataFunc = nullptr);
+
+    // Execute all collectives on all test children
+    // Blocks until collective is completed
+    void ExecuteCollectives();
+
+    // Perform results validation - compare output to expected
+    void ValidateResults(bool& isCorrect, int collId = -1, int const rank = -1);
+
+    // Release allocated memory
+    void DeallocateMem(int collId = -1, int const rank = -1);
```

Comment:
might make sense to be grouped with AllocateMem Function

Reply:
Currently this is ordered in the general flow through a collective call.  Deallocation of memory generally occurs after results have been validated.